### PR TITLE
Add support for trio and curio in Python async.

### PR DESCRIPTION
**Pull request by miracle2k**
There are two alternative async libraries for Python, 
curio (https://github.com/dabeaz/curio) and trio 
(https://github.com/python-trio/trio). They are reasonably
popular, and improve on asyncio as being being 
async/await all the way down, while asyncio is internally
based on callbacks.

I wanted to add support via a separate import first, i.e.
"import ccxt.trio as ccxt", but it turns out we really only
have to switch out aiohttp, so I think it makes more sense
to just use an argument. The http library used is "asks",
which the user has to install themselves if they want 
trio/curio support. In the future, it seems urllib3 will 
support trio and curio, so at that point we can switch
(https://github.com/njsmith/urllib3/issues/1).

I did not yet port throttle.py. I wanted to get some feedback
first.

Usage is like this:

```
import curio
import ccxt.async as ccxt

k = ccxt.kraken({'async_mode': 'curio'})

async def load():
    m = await k.load_markets()
    print(len(m))

curio.run(load())
```

### CODE REVIEW

**Comment by kroitor** [comment]
Hi! ) I think, a better way to use trio and curio interchangeably is to not add an `asyncio_mode` option, and, instead, add an `asyncio` option.

So, the usage would be:

```Python
import curio
import ccxt.async as ccxt
k = ccxt.kraken ({'asyncio': curio})
```

It is much cleaner that way and does not require any dependencies to be added to ccxt at all. 

I'm thinking of the above because I barely see a difference between these two methods:

```Python
    async def _fetch_with_asyncio(self, method, url, headers, encoded_body, body):
        session_method = getattr(self.session, method.lower())
        http_status_code = None

        try:
            async with session_method(yarl.URL(url, encoded=True),
                                      data=encoded_body,
                                      headers=headers,
                                      timeout=(self.timeout / 1000),
                                      proxy=self.aiohttp_proxy) as response:
                http_status_code = response.status
                text = await response.text()
                self.last_http_response = text
                self.last_response_headers = response.headers
                self.handle_errors(http_status_code, text, url, method, self.last_response_headers, text)
                self.handle_rest_errors(None, http_status_code, text, url, method)
                if self.verbose:
                    print("\nResponse:", method, url, str(http_status_code), str(response.headers), self.last_http_response)
                self.logger.debug("%s %s, Response: %s %s %s", method, url, response.status, response.headers, self.last_http_response)

        except socket.gaierror as e:
            self.raise_error(ExchangeNotAvailable, url, method, e, None)

        except concurrent.futures._base.TimeoutError as e:
            self.raise_error(RequestTimeout, method, url, e, None)

        except aiohttp.client_exceptions.ClientError as e:
            self.raise_error(ExchangeError, url, method, e, None)

        self.handle_errors(http_status_code, text, url, method, self.last_response_headers, text)
        return self.handle_rest_response(text, url, method, headers, body)

    async def _fetch_with_asks(self, method, url, headers, encoded_body, body):
        session_method = getattr(self.session, method.lower())
        http_status_code = None

        try:
            response = await session_method(url,
                                            data=encoded_body,
                                            headers=headers,
                                            timeout=(self.timeout / 1000))
            http_status_code = response.status_code
            text = response.text
            self.last_http_response = text
            self.last_response_headers = response.headers
            self.handle_errors(http_status_code, text, url, method, self.last_response_headers, text)
            self.handle_rest_errors(None, http_status_code, text, url, method)
            if self.verbose:
                print("\nResponse:", method, url, str(http_status_code), str(response.headers), self.last_http_response)
            self.logger.debug("%s %s, Response: %s %s %s", method, url, response.status_code, response.headers, self.last_http_response)

        except socket.gaierror as e:
            self.raise_error(ExchangeNotAvailable, url, method, e, None)

        except concurrent.futures._base.TimeoutError as e:
            self.raise_error(RequestTimeout, method, url, e, None)

        self.handle_errors(http_status_code, text, url, method, self.last_response_headers, text)
        return self.handle_rest_response(text, url, method, headers, body)
```

↑ This looks like one method to me, but, duplicated without the need for duplication. If we just switch all `asyncio` calls in Exchange.js to `self.asyncio` it would resolve the issue very smoothly, wouldn't it?

Let me know, what do you think about it. Thx!

**Comment by miracle2k** [comment]
That wouldn't work :/ The problem is that `aiohttp`, used for http requests, does not support curio/trio; the two methods do look similar, but the new one uses a new library `asks`. Although it looks like duplicated code, it is distinct but looks similar because the API of both libraries is quite similar (but notably not the same - note `response.status_code` vs `response.status`, or `await response.text()` vs `response.text`.

In terms of adding a dependency, I suggest that we do *not* add one. If anyone wants to use trio or curio, they should install it themselves, similar to a peer dependency in JavaScript. Most users do not need/use trio, and we shouldn't force it on them.

Having said that, in Python, `setup.py` supports a mechanism by which the installer could run `pip install ccxt[curio]` to install the "curio" version with optional curio dependencies. However, I would go as far as saying that we do not have to support this either, and we do not have to concern ourselves with one more dependency. People who want to use trio will need to have it installed anyway.

-----

To give a further explanation about why we need a separate http library: It is not a question of switching out one module API for another with the same API. 

The way this works, in Python, is that if anywhere in the code you have ``x = await foo``, then `foo` is given to the event loop; it is either given to ``curio.run()``, or to `asyncio.get_event_loop().run_forever()`. They will then handle `foo` (and internal "trap" object) and continue the execution once the result `x` is ready.

So if, in ccxt, we do ``await aiohttp.get()``, which might call ``await asyncio.socket_read()``, then the return value of ``asyncio.socket_read()``, which is a trap object specific to `asyncio` will be given to the loop running via ``curio.run``, which cannot handle asyncio traps.

In other works, this will not work:

```
async def test():
    await asyncio.sleep(1)

curio.run(load())
```

and neither will this:

```
async def fake_aiohttp_get():
     return await asyncio.open_connection()

async def test():
    await fake_aiohttp_get('http://kraken.com/api')

curio.run(load())
```

**Comment by kroitor** [comment]
> it is distinct but looks similar because the API of both libraries is quite similar (but notably not the same - note response.status_code vs response.status, or await response.text() vs response.text.

I literally mean, 90% of the other code is the same, so no need to duplicate it... I've noticed the differences in committed code, but I mean, when 2 of 20 lines differ, you don't really want to copy all 20.

> In terms of adding a dependency, I suggest that we do not add one.

Yep, but you did add one. The `asks` package. You made it optional, but it doesn't solve the issue with adding a dependency even when it's optional. Because the issue with dependencies is that you have to check for security and monitor their code, as discussed here: https://github.com/ccxt/ccxt/issues/943

Basically, I would go without asks, and I can clearly see a way to do it without adding a dependency, while still supporting it.

> So if, in ccxt, we do await aiohttp.get(),...

We will do self.aiohttp.get (), and make both the loop (`asyncio`) and the http-client (`aiohttp`) optional and overrideable. Need just a little time to lay it out, among other tasks.

**Comment by miracle2k** [comment]
I like the idea of simply passing in two replacement modules; it would keep the code in ccxt simpler. But I don't see how that would play out. The API surface of aiohttp that ccxt uses is fairly complex large - creating a session, a connector and so on.

If I understand you correctly, my code calling into ccxt would then need to look like this:

```
import curio
import asks_wrapper
import ccxt.async as ccxt
k = ccxt.kraken ({'asyncio': curio, 'aiohttp': asks_wrapper})
```

```
# asks_wrapper.py

import asks

class TCPConnector(object):
    def __init__(self, *a, **kw):
         pass

class ClientSession(object):
     def __init__(self, *a, **kw):
          self._session = asks.Session(....)

     async def get(self, *a, **kw):
          response = await self._session.get(....)
          return AiohttpStyleResponse(response)

class AiohttpStyleResponse(object):
     """'
     Rewrite status_code, text() and so on...
     """
```
Is that the idea?

> Need just a little time to lay it out with some very basic introspection, among other tasks.

What do you mean by that?

**Comment by kroitor** [comment]
> Is that the idea?

> What do you mean by that?

It's actually easier to be done, than said... ;)

Let look at your changes scrupulously (i'm skipping the minor and *"obvious"* ones, and I only compare the most significant or *"breaking"* changes).

=================

1. This snippet:

```Python
        if async_mode == 'asyncio':
            if 'asyncio_loop' in config:
                self.asyncio_loop = config['asyncio_loop']
            self.asyncio_loop = self.asyncio_loop or asyncio.get_event_loop()

            self.own_session = 'session' not in config
            if self.own_session:
                # Create out SSL context object with our CA cert file
                context = ssl.create_default_context(cafile=certifi.where())
                # Pass this SSL context to aiohttp and create a TCPConnector
                connector = aiohttp.TCPConnector(ssl_context=context, loop=self.asyncio_loop)
                self.session = aiohttp.ClientSession(loop=self.asyncio_loop, connector=connector)

        # ---------------- THE ADDED CODE IS MOSTLY BELOW: -------------------------

        elif async_mode == 'trio' or async_mode == 'curio':
            if not asks:
                raise ValueError('async_mode "%s" requires the "asks" package to be installed.' % async_mode)
            asks.init(async_mode)

            self.own_session = 'session' not in config
            if self.own_session:
                self.session = asks.Session(connections=4)
```

This added snippet of code is not needed. Why? Because you could just create the  outside of ccxt and pass it into the constructor:

```Python
ccxt.kraken({'session': asks.Session(connections=4)})
```

↑ The above short single line effectively does the same as you do in that snippet of added code. Well, almost, but, let's assume we do it this way and continue further with the reasoning, just to see what comes out of it...

==================

2. I'm not mentioning all the added if-checks in the descructor and `close()`, for obvious reasons... The next *"noticeable"* change is:

```Python
        if self.async_mode == 'asyncio':
            return await self._fetch_with_asyncio(method, url, headers, encoded_body, body)
        elif self.async_mode == 'trio' or self.async_mode == 'curio':
            return await self._fetch_with_asks(method, url, headers, encoded_body, body)
        else:
            raise ValueError('Not a supported async mode string: %s' % self.async_mode)
```

I'd say, it should really be:

```Python
    return await self.just_fetch(...all the same args...)  # nothing else needed, the secret is inside of it
```

So, again, looks like the second added change is not needed and we get over the issue (at this point) with the code that... uhm... pretty much existed already ) You may disagree, but, in general, for now we've shown that with point 1 and 2 the changes could be done in a much easier way. Please, follow me to the next point )

==================

3. This is the most significant change, and we need to compare two methods line-by-line to understand and see that the difference needed to activate support for trio and curio is very-very small, and current ccxt already has in it practically 99% of all the stuff needed to do that.

```Python
    async def _fetch_with_asyncio(self, method, url, headers, encoded_body, body):
        session_method = getattr(self.session, method.lower())
        http_status_code = None

        try:
            async with session_method(yarl.URL(url, encoded=True),
                                      data=encoded_body,
                                      headers=headers,
                                      timeout=(self.timeout / 1000),
                                      proxy=self.aiohttp_proxy) as response:
                http_status_code = response.status
                text = await response.text()
                self.last_http_response = text
                self.last_response_headers = response.headers
                self.handle_errors(http_status_code, text, url, method, self.last_response_headers, text)
                self.handle_rest_errors(None, http_status_code, text, url, method)
                if self.verbose:
                    print("\nResponse:", method, url, str(http_status_code), str(response.headers), self.last_http_response)
                self.logger.debug("%s %s, Response: %s %s %s", method, url, response.status, response.headers, self.last_http_response)

        except socket.gaierror as e:
            self.raise_error(ExchangeNotAvailable, url, method, e, None)

        except concurrent.futures._base.TimeoutError as e:
            self.raise_error(RequestTimeout, method, url, e, None)

        except aiohttp.client_exceptions.ClientError as e:
            self.raise_error(ExchangeError, url, method, e, None)

        self.handle_errors(http_status_code, text, url, method, self.last_response_headers, text)
        return self.handle_rest_response(text, url, method, headers, body)

    # ---------------- THE ADDED CODE IS MOSTLY BELOW: -------------------------

    async def _fetch_with_asks(self, method, url, headers, encoded_body, body):
        session_method = getattr(self.session, method.lower())
        http_status_code = None

        try:
            response = await session_method(url,
                                            data=encoded_body,
                                            headers=headers,
                                            timeout=(self.timeout / 1000))
            http_status_code = response.status_code
            text = response.text
            self.last_http_response = text
            self.last_response_headers = response.headers
            self.handle_errors(http_status_code, text, url, method, self.last_response_headers, text)
            self.handle_rest_errors(None, http_status_code, text, url, method)
            if self.verbose:
                print("\nResponse:", method, url, str(http_status_code), str(response.headers), self.last_http_response)
            self.logger.debug("%s %s, Response: %s %s %s", method, url, response.status_code, response.headers, self.last_http_response)

        except socket.gaierror as e:
            self.raise_error(ExchangeNotAvailable, url, method, e, None)

        except concurrent.futures._base.TimeoutError as e:
            self.raise_error(RequestTimeout, method, url, e, None)

        self.handle_errors(http_status_code, text, url, method, self.last_response_headers, text)
        return self.handle_rest_response(text, url, method, headers, body)
```

Ok, this one is a little bigger due to duplication, let's compare:

----------------------------

3a. 

Existing code:

    ```Python
    async def _fetch_with_asyncio(self, method, url, headers, encoded_body, body):
        session_method = getattr(self.session, method.lower())
        http_status_code = None

        try:
            async with session_method(yarl.URL(url, encoded=True),
                                      data=encoded_body,
                                      headers=headers,
                                      timeout=(self.timeout / 1000),
                                      proxy=self.aiohttp_proxy) as response:
    ```

Added code:

    ```Python
    async def _fetch_with_asks(self, method, url, headers, encoded_body, body):
        session_method = getattr(self.session, method.lower())
        http_status_code = None

        try:
            response = await session_method(url,
                                            data=encoded_body,
                                            headers=headers,
                                            timeout=(self.timeout / 1000))
    ```

↑ to me it looks like the only line that is different there is the proxy line (the last one), which can easily be made optional, I hope I don't need to explain, how, as it's pretty much obvious, again...

--------------------------

3b. The second part of the comparison, you said there will be a problem because of the difference shown here:

Existing code:

    ```Python
            http_status_code = response.status
            text = await response.text()
    ```

Added code:

    ```Python
            http_status_code = response.status_code
            text = response.text
    ```

Here, again, we can just look into `response` and see if `.status_code` or `.status` is there, I guess, I don't have to show how... The await can be added optionally if `.text` is something awaitable, again... So, once more, to me the above two lines look like the same lines and they don't have to be separated, just need a very slight rewrite to account for the mentioned differences, and that's it. So, in general, same code again, and the so-massive change isn't really needed here as well.

----------------------------

3c. And the final part of the comparison is the rest of the two methods, and, again, I hope I don't need to copypaste it once more just to show that the entire resting parts of the two methods are *"practically identical"*...

Aparta from that, I don't see anything else that should be done in order to add support for trio, curio and asks without actually adding any redundant code ) Isn't it beautiful? We can support them without even adding their package into ccxt ) 

=================================+

> If I understand you correctly, my code calling into ccxt would then need to look like this:

> ```Python
> import curio
> import asks_wrapper
> import ccxt.async as ccxt
> k = ccxt.kraken ({'asyncio': curio, 'aiohttp': asks_wrapper})
> ```

Well, almost, but I would make it even easier for you:

```Python
import asks
import ccxt.async as ccxt
k = ccxt.kraken ({'session': asks.Session(connections=4)})
```

↑ Isn't this (existing) option an elegant solution to the entire issue?

Let me know, what you think of it... Thx!

**Comment by miracle2k** [comment]
First, let me thank you for taking so much time for this issue. I really appreciate it.

I think I see what you are going for here, and I'll try to create a patch that implements this procedure. I will say for the record that I disagree with the approach; I think we now have ccxt testing for various variants of the `Session` and `Response` interfaces, instead of a two clearly defined ones. To be fair, I think I could have removed *some* of the duplication in the fetch method by moving them to a shared helper.

However, it is your library, and I'd really like to see support for trio in it, so I will try to rewrite the pull request.

**Comment by kroitor** [comment]
@miracle2k thx for your understanding, I would go with as little footprint as possible when it comes to dependencies. And I also really appreciate your participation in this, it is very important to me, and to the community as well! Hope to see more from you ;)

**Comment by miracle2k** [comment]
This is how it would need to look to support trio. Note that ``session.get`` in ``aiohttp`` returns a context manager, but in ``asks`` returns just a response object, so we need to test for ``__aenter__``.

Personally, I think this is the wrong approach. It might be best to wait, since it seems ``requests`` will gain async support in the future via ``urllib3`` and it seem like it will handle trio (https://github.com/kennethreitz/requests-core), which might allow you to drop ``aiohttp`` completely and just use ``requests``.

**Comment by miracle2k** [comment]
For now, this is the hack that I am using:

```python
from ccxt.async.base.exchange import Exchange

class TrioExchange(Exchange):
    
    async def fetch(self, url, method='GET', headers=None, body=None):
        headers = self.prepare_request_headers(headers)

        url = self.proxy + url

        if self.verbose:
            print("\nRequest:", method, url, headers, body)

        self.logger.debug("%s %s, Request: %s %s", method, url, headers, body)

        encoded_body = body.encode() if body else None

        session_method = getattr(asks, method.lower())
        http_status_code = None

        try:
            response = await session_method(str(yarl.URL(url, encoded=True)),
                                      data=encoded_body,
                                      headers=headers,
                                      timeout=(self.timeout / 1000))
            http_status_code = response.status_code
            text = response.text
            self.last_http_response = text
            self.last_response_headers = response.headers
            self.handle_errors(http_status_code, text, url, method, self.last_response_headers, text)
            self.handle_rest_errors(None, http_status_code, text, url, method)
            if self.verbose:
                print("\nResponse:", method, url, str(http_status_code), str(response.headers), self.last_http_response)
            self.logger.debug("%s %s, Response: %s %s %s", method, url, response.status_code, response.headers, self.last_http_response)

        except socket.gaierror as e:
            self.raise_error(ExchangeNotAvailable, url, method, e, None)

        except concurrent.futures._base.TimeoutError as e:
            self.raise_error(RequestTimeout, method, url, e, None)

        except aiohttp.client_exceptions.ClientError as e:
            self.raise_error(ExchangeError, url, method, e, None)

        self.handle_errors(http_status_code, text, url, method, self.last_response_headers, text)
        return self.handle_rest_response(text, url, method, headers, body)

async def main():
    k = ccxt.kraken()
    k.__class__.__bases__ = (TrioExchange,)
    print(await k.fetch_markets())
    await k.close()
```

❤️ Python.

I am interpreting @kroitor 's thumbs up as an agreement that we do not go with this approach for now and instead wait for support in urllib3/requests.
